tools/update_roles.php

```diff
@@ -0,0 +1,744 @@
+}
+
+/**
+ * UserHasPermission
+ *
+ * Checks if a given user has a given permission
+ *
+ * @param int    $userID     the user ID
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function userHasPermission($userID, $permission)
+{
+    $DB = Database::singleton();
+
+    $permission = $DB->pselect(
+        "SELECT upr.userID
+         FROM user_perm_rel upr
+         LEFT JOIN permissions p USING (permID)
+         WHERE upr.userID=:userID AND p.code=:permission",
+        array(
+         'userID'     => $userID,
+         'permission' => $permission,
+        )
+    );
+
+    if (empty($permission)) {
+        return false;
+    }
+    return true;
+}
+
+/**
+ * RoleHasPermission
+ *
+ * Checks if a given role contains a given permission
+ *
+ * @param string $role       the name of the role
+ * @param string $permission the name of the permission
+ *
+ * @return boolean
+ */
+function roleHasPermission($role, $permission)
+{
+    $rolePermissions = getRolePermissions($role);
+
+    if (in_array($permission, $rolePermissions)) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * RefreshRoles
+ *
+ * Rebuilds roles based on user permissions
+ *
+ * @return boolean
+ */
+function refreshRoles()
+{
+    $DB = Database::singleton();
+
+    echo "Rebuilding roles based on user permissions\n\n";
+
+    // Iterate over each user, update their roles based on their permissions
+    foreach (getUserIDs() as $user) {
+
+        $userID = $user['ID'];
+
+        echo "Updating roles for ${user['Real_name']}\n";
+
+        // delete existing roles
+        $DB->delete(
+            'users_permission_categories_rel',
+            array('user_id' => $userID)
+        );
+
+        // recalculate their roles
+        $newRoles = getUserRolesByPermission($user['ID']);
+
+        // update their roles in the database
+        foreach ($newRoles as $role) {
+            echo "\tAdding $role\n";
+            $DB->insert(
+                'users_permission_categories_rel',
+                array(
+                 'user_id'                => $userID,
+                 'permission_category_id' => getRoleID($role),
+                )
+            );
+        }
+
+        echo "\n";
+    }
+}
+
+/**
+ * RefreshPermissions
+ *
+ * Tops up user permissions based on their roles
+ * Adds any permissions a user may be missing from a role they are assigned
+ *
+ * @return boolean
+ */
+function refreshPermissions()
+{
+    $DB = Database::singleton();
+
+    echo "Updating permissions based on user roles\n\n";
+
+    // Iterate over each user, update their roles based on their permissions
+    foreach (getUserIDs() as $user) {
+
+        echo "Updating permissions for ${user['Real_name']}\n";
+
+        // get user roles
+        $roles = getUserRoles($user['ID']);
+
+        // update their permissions in the database
+        foreach ($roles as $role) {
+            $permissions = getRolePermissions($role);
```

Comment:
this fails because you are passing an array and then querying on it instead of getting the ID from that array

$role is an array,
you want to pass $role['category'] which is not currently in the array passed